### ISSUE TITLE
Attempt to automatically deduplicate publication metadata after incoming DOI has been confirmed

### ISSUE THREAD

**Issue by EricDurante** [psu-libraries/researcher-metadata#630]
Publications with file uploads from Activity Insight that are entering the new open access deposit workflow may already have duplicate metadata in RMD that was imported from another record in Activity Insight or from a different source altogether. Some of this metadata could be useful later on in this workflow, and we'll ultimately want it to be deduplicated in RMD anyway. Having a correct/confirmed DOI for the incoming publication will greatly facilitate any automated or manual deduplication that needs to be done. With that being the case, it makes sense that the next step in this workflow after DOI verification should be metadata deduplication.

This could work as follows:
1. The result of either [automated](https://github.com/psu-libraries/researcher-metadata/issues/626) or [manual](https://github.com/psu-libraries/researcher-metadata/issues/629) DOI verification will be that a new deduplication background job is enqueued.
2. The background job will first find and group and apparent duplicates of the incoming publication using existing RMD functionality.
3. The background job will then attempt to automatically deduplicate the metadata by DOI (if available) using existing RMD functionality.
4. If not every possible duplicate has a DOI, then we may try to automatically deduplicate based on other metadata. (I'm not sure that we'll want to do this)
5. If all duplication can be resolved automatically, then the current job can enqueue another background job to handle the next step in the workflow (open access status check).
6. If not all duplication can be resolved automatically, then the job will exit, leaving the possible duplicates grouped for manual deduplication.

**Comment by EricDurante** [comment]
It occurs to me that we could potentially run into a problem here given that we're going to end up with multiple metadata deduplication workflows in RMD (this being one of them).

The idea here is that after we attempt this automatic step of duplicate detection and deduplication, we'll be left with some possible duplicates that can't be deduplicated automatically. These will end up in another [to-do](https://github.com/psu-libraries/researcher-metadata/issues/677) list in this new workflow, and RMD admins will ultimately be able to take some action from this list that will manually resolve the duplication and move the publication onward in the workflow.

The trouble that I can see is that if we show publications in this list based only on the presence of an associated file upload and on the presence of possible duplicate metadata, then publications may disappear from the list if they happen to be deduplicated through one of the other existing workflows. If this were to happen, then the publication wouldn't be pushed forward in this workflow upon completion of its manual deduplication step because its manual deduplication step would never happen. This is leading me to think that we may need to track the status of each publication in this workflow more explicitly. 

This is just one example of how publications could potentially "fall out" of this workflow if we implement it as a series of automated, linked steps where each step is either triggered by the previous one (if the previous step is successful) or by a manual action (if the previous step is not successful). At each pause for manual action (including the DOI verification step before this), a publication could potentially get dropped from this workflow if some other existing (or future) process just happens to do what needs to be done at that step first.

I was hoping to avoid tracking workflow state like this, but it might be unavoidable. So what I'm thinking is that for each publication going through the workflow we'd have to store its status whenever it requires manual action (i.e. "requires_manual_deduplication"), update its status at the completion of the manual step, and present it to RMD admins for manual action based solely on that status. So this could sometimes leave publications queued up for manual action that no longer require any action other than moving them on to the next step. That's a little awkward, so if it turned out to be a frequent occurrence, then maybe we'd want to implement an automated process that checks for publications that can be automatically advanced in the workflow and pushes them forward.

@ajkiessl Do you see what I'm saying here? I'm not sure how well I explained it..

**Comment by EricDurante** [comment]
@ajkiessl Yep - what you're suggest occurred to me too, but I really don't think that I'd want other areas of the code to need to know about this workflow for the exact reasons that you already mentioned. I think that we can leave it entirely up to the manual admin actions that occur in this workflow to change the workflow state. Some other process might end up merging the duplicates, leaving no work for the admin to actually do at this step in this workflow, but they'd still have to press the button to move the publication to the next step. 

That's the bit that's a little awkward in terms of this workflow's UI, but it allows the merging logic to remain completely unaware of this workflow and its state. It's also something that could be automated if necessary. So if we ended up with a pile of publications with the status "requires_manual_deduplication" that were then deduplicated via the existing manual deduplication workflow, we could have a cron job that checks for publications in that state. If it found a publication with the status "requires_manual_deduplication" that didn't belong to a duplicate group, then it could be responsible for updating the status and kicking off the next background job instead of waiting on the admin the push the button. 

I know that's another piece of machinery in the mix, but I think it keeps this workflow self-contained while streamlining it as much as possible. I think it also has the side benefit of allowing us to leverage more of the existing RMD admin functionality in this new workflow (at least temporarily, if we choose). Instead of implementing a new deduplication UI that's tailored for this workflow, we _could_ send the admin user to the existing rails_admin deduplication UI to do the actual work, and then require them to come back to the new workflow to move the publication to the next step - or we implement the cron job that does that part for them automatically.

What do you think?

**Comment by EricDurante** [comment]
Something else that we're going to have to consider is that new duplicates of a given publication could potentially show up at any point in the workflow as the result of a co-author adding the same publication (with file upload) to Activity Insight. So even after a publication has been through the deduplication step in this workflow once, it might get grouped with a new incoming duplicate, and it may either get merged into the new one (along with its file upload), or the new one (including its file upload) may be merged into the original. That's definitely a problem for our current concept of a nice, linear workflow with each step linked in series.

In the scenario above, if the original got merged into a new, incoming duplicate, then the normal, linear workflow would be fine, I think. But if the new duplicate got merged into the original that was already further along in the process, then we'd need to have a check at some point later in the process to ensure that we didn't have more than one file for that publication prior to deposit. If we did find more than one file, then we'd have to send the publication back to the file deduplication step first.

Also if a new duplicate showed up _without_ a file upload, and the original (with an upload) got merged into it, then I think the publication would drop out of this workflow entirely. I don't know how we'd avoid that other than having a separate process that would check for this situation and push these publications back into the workflow.

I feel like there are almost surely other issues like this that I haven't thought of yet. The complexity of the solution makes me wonder if there's a better overall way to design it.

**Comment by ajkiessl** [comment]
We could make deduplication the first step as a cronjob and any publication can't have it's DOI checked until it is deduplicated.  Then, the DOI check could be another regular cronjob.  In fact, we could potentially make all the automated steps cronjobs that only act on publications that fit the criteria for that job.  We could then filter out anything that is part of a duplicate group for the stages past the deduplication stage.  Anytime a publication that is along its way in the OA Workflow ends up in a duplicate group again, it would show back up in the first step to be deduplicated.  If the deduplication results in a merge, then it resets all the checks and the statuses that indicate it needs manual work, and it will once again be picked up by the DOI cron job.  If it needs manual merging and an admin decides they are not duplicates, then it will just return to the last spot it was in the workflow (the scope will pick it back up).  

If any of the automated cronjob steps fail (ie manual verification is now needed), the publication could be flagged like you mentioned before (something like "requires_manual_doi_check").  The dashboard could then use the same scopes as the cronjobs would use, but also require this flag.  That way, if the publication is grouped while waiting for manual action, it will show back up in the first step's scope.

Basically, what I described above will reset the workflow anytime there is a merge.  That might be too broad, but it may be the easiest way to handle this initially.

**Comment by ajkiessl** [comment]
We may not need to make _everything_ a cronjob, but the DOI check should be, to kick things off.  Everything else can be linked one after another after the last job/manual action.

**Comment by ajkiessl** [comment]
The autodedupe would be a cronjob too, but it wouldn't really be triggering anything after it.

**Comment by EricDurante** [comment]
I was kicking around an idea somewhat like this also. I think it's worth exploring.

**Comment by ajkiessl** [comment]
We've decided we are not concerning ourselves with deduplication as a step and requirement to proceed through the workflow.  For the most part, deduplication is not necessary for this workflow to work.  Closing this issue.
